internal/oci/container.go

```diff
@@ -349,18 +392,110 @@ func (c *Container) exitFilePath() string {
-	process, err := os.FindProcess(c.state.Pid)
+	_, err := c.pid()
 	if err != nil {
-		logrus.Infof("container %s not found: %v", c.id, err)
+		logrus.Errorf("checking if pid of %s is running failed: %v", c.id, err)
 		return false
 	}
 
-	if err := process.Signal(syscall.Signal(0)); err != nil {
-		logrus.Errorf("container %s not running %v", c.id, err)
-		return false
+	return true
+}
+
+// Pid is the external, locking function that returns the container's
+// pid, if it's running, and actually the container process
+func (c *Container) Pid() (int, error) {
+	c.opLock.Lock()
+	defer c.opLock.Unlock()
+	return c.pid()
+}
+
+// pid is the internal, non-locking version of Pid. It checks
+// that we have a Pid defined in the state, that pid can be found
+// and it is the same process that was originally started by the runtime
+func (c *Container) pid() (int, error) {
+	if c.state == nil {
+		return -1, errors.New("state not initialized")
+	}
+	pid := c.state.InitialPid
+	if pid <= 0 {
+		return -1, errors.New("pid not initialized")
 	}
 
-	return true
+	// container has stopped (as pid is initialized but the runc state has overwritten it)
+	if c.state.Pid == 0 {
+		return 0, findprocess.ErrNotFound
+	}
+
+	if _, err := findprocess.FindProcess(pid); err != nil {
+		return -1, err
+	}
+
+	if err := c.verifyPid(pid); err != nil {
+		return -1, err
+	}
+	return pid, nil
+}
+
+// findAndReleasePid attempts to find the container's pid
+// and if it's found, release it. In between these steps,
+// it also verifies the pid running is the same as was originally
+// started by the runtime
+func (c *Container) findAndReleasePid() (bool, error) {
+	pid := c.state.InitialPid
+
+	if err := c.verifyPid(pid); err != nil {
+		if !errors.Is(err, errContainerPidStopped) {
+			return false, err
+		}
+		return false, nil
+	}
+
+	process, err := findprocess.FindProcess(pid)
+	if err == findprocess.ErrNotFound {
+		return false, nil
+	}
+	if err != nil {
+		return false, err
+	}
+
+	// the process was found and was the correct one
+	// release it
+	if err := process.Release(); err != nil {
+		return true, err
+	}
+
+	return true, nil
+}
+
+// verifyPid checks that the start time for the process on the node is the same
+// as the start time we saved after creating the container.
+// This is the simplest way to verify we are operating on the container
+// process, and haven't run into pid wrap
+func (c *Container) verifyPid(pid int) error {
+	startTime, err := getProcessStartTime(pid)
+	if err != nil {
+		return err
+	}
+
+	if startTime != c.state.ProcessStartTime {
+		return errors.New("pid running but not the original container. pid wrap may have occurred")
+	}
+	return nil
+}
+
+// getProcessStartTime reads the kernel's /proc entry for stime for pid
+func getProcessStartTime(pid int) (string, error) {
+	st, err := os.Stat(fmt.Sprintf("/proc/%d", pid))
+	if err != nil {
+		return "", errContainerPidStopped
+	}
+
+	sysInfo, ok := st.Sys().(*syscall.Stat_t)
+	if !ok {
```

Comment:
this is a no-op as os.Stat() always return `syscall.Stat_t`.

You can avoid all this checking if you use `unix.Stat` instead.

Reply:
fixed